### PR TITLE
Improve error reporting

### LINKED ISSUE #38: Improve error reporting for capturing regular vars in modal contexts
Right now one gets
```
$ mtt infer -e "λf:B -> []A. λy:B. (λx:[]A. box x) (f y)"
Type inference error: Variable x is not found in the regular context
```
which is not particularly helpful and might be even misleading.

### PR DESCRIPTION
**Pull request by GoPavel**
- [X] introduce typed errors (by polymorphic variant)
- [x] fix #38 (according to discussion)
- [X] refactor Typechecker

### CODE REVIEW

**Comment by GoPavel** [comment]
@anton-trunov Could you prereview changes before format problems will be fixed?

File: src/Typechecker.ml [hunk 1]
```diff
@@ -3,93 +3,97 @@ open Result.Let_syntax
 open Ast
 open Ast.Expr
 
-type error = string
+type error = [
+  | `TypeMissmatch of string
```
Comment (anton-trunov) [inline src/Typechecker.ml]:
```suggestion
  | `TypeMismatch of string
```

File: src/Typechecker.ml [hunk 2]
```diff
@@ -3,93 +3,97 @@ open Result.Let_syntax
...
+  | `UnboundRegularVarInsideBox of Location.t (* var loc *)
+                                   * string (* message *)
```
Comment (anton-trunov) [inline src/Typechecker.ml]: If you need to add comments for components then you probably want an anonymous record here.

File: src/Env.ml [hunk 3]
```diff
@@ -1,6 +1,10 @@
 open Base
 open Result.Let_syntax
 
+type error = [ `EnvUnboundVariable of string (* name of variable *)
+                                      * string (* message *)
+             ]
```
Comment (anton-trunov) [inline src/Env.ml]: Perhaps anonymous records would be nicer than comments.

File: src/Env.ml [hunk 4]
```diff
@@ -14,7 +18,7 @@ module Make (Key : sig
   [@@@end]
 
   include Equal.S with type t := t
-end) =
+end) (EnvCfg: sig val envName : string end) =
```
Comment (anton-trunov) [inline src/Env.ml]: Maybe instead of adding this parameter to the functor we can move it to `ID` signature.

File: src/Location.ml [hunk 5]
```diff
@@ -29,8 +29,22 @@ let locate_start_end data s_pos e_pos =
   let loc = mk s_pos e_pos in
   { data; loc }
 
+(* TODO: maybe replace locate with optional [loc] with
+         locate: t -> a -> a located
+         no_location: a -> a located *)
```
Comment (anton-trunov) [inline src/Location.ml]: Let's create an issue for this.

**Review by anton-trunov — APPROVED** [review]

File: src/Env.ml (hunk 3 above)
Comment (GoPavel) [reply src/Env.ml]: It seems like Ocaml cannot parse that.
I can't google that limitation, maybe it is implied Ocaml's grammar, but the following code doesn't compile.
```ocaml
type error = [`Error of { msg : string} ]
```

Comment (GoPavel) [reply src/Env.ml]: Explicit type like `type SomeErrorArg = { msg: string }` looks awful and too verbose. It helps with constructing and destructing, but the type of error will become silly. (`'SomeError SomeErrorArgs`) .
So I would keep the current version with comments.

File: src/Typechecker.ml (hunk 1 above)
Comment (GoPavel) [reply src/Typechecker.ml]: fixed

File: src/Env.ml (hunk 4 above)
Comment (GoPavel) [reply src/Env.ml]: done

File: src/Typechecker.ml (hunk 2 above)
Comment (GoPavel) [reply src/Typechecker.ml]: I replied in another thread.

File: src/Env.ml (hunk 3 above)
Comment (anton-trunov) [reply src/Env.ml]: It's my bad: polymorphic variants cannot be combined with inline records, see https://stackoverflow.com/questions/61503391/inline-records-in-polymorphic-variants. Ok, let's see how it goes as it is -- we can always refactor it if there is a need.

File: src/Id.ml [hunk 6]
```diff
@@ -8,6 +8,8 @@ module type ID = sig
 
   val to_string : t -> string
 
+  val contextName : string
```
Comment (anton-trunov) [inline src/Id.ml]:
```suggestion
  val context_kind : string
```

File: src/Location.ml (hunk 5 above)
Comment (GoPavel) [reply src/Location.ml]: done: #53

File: src/Id.ml (hunk 6 above)
Comment (GoPavel) [reply src/Id.ml]: fixed

File: bin/mtt.ml [hunk 7]
```diff
@@ -4,6 +4,19 @@ open Result.Let_syntax
...
 open ParserInterface
 
+let formatLError Location.{ data = error; loc } =
+  match error with
+  | `TypeMismatchError msg -> Location.pp ~msg loc
+  | `EvaluationError msg -> Location.pp ~msg loc
+  | `EnvUnboundVariableError (_, msg) -> Location.pp ~msg loc
+  | `UnboundRegularVarInsideBoxError (_, msg) -> Location.pp ~msg loc
+
+let formatError error =
```
Comment (anton-trunov) [inline bin/mtt.ml]:
```suggestion
let format_error_no_loc error =
```

File: bin/mtt.ml [hunk 8]
```diff
@@ -4,6 +4,19 @@ open Result.Let_syntax
 open Mtt
 open ParserInterface
 
+let formatLError Location.{ data = error; loc } =
```
Comment (anton-trunov) [inline bin/mtt.ml]: Sorry, I missed the following two identifiers, let's use more idiomatic names here.
Also, error messages should have location information by default, hence I suggest we change it as follows:
```suggestion
let format_error Location.{ data = error; loc } =
```

**Review by anton-trunov — CHANGES_REQUESTED** [review]

File: bin/mtt.ml (hunk 8 above)
Comment (GoPavel) [reply bin/mtt.ml]: Ok, thank you. I forgot to check naming too after the first review.

**Review by anton-trunov — APPROVED** [review]

**Pull request merged**